--- check_rbl-1.1.0/check_rbl~	2010-02-01 18:27:44.918315737 +0200
+++ check_rbl-1.1.0/check_rbl	2010-02-01 18:28:07.075134119 +0200
@@ -272,7 +275,7 @@
 my $total = scalar @BLACKLISTED;
 
 my $status =
-  "BLACKLISTED on $total " . ( ( $total == 1 ) ? 'server' : 'servers' );
+  $OPTIONS->host. " BLACKLISTED on $total " . ( ( $total == 1 ) ? 'server' : 'servers' );
 
 if ( $total > 0 ) {
     $status .= " (@BLACKLISTED)";
